### PR TITLE
add decimal support

### PR DESCRIPTION
**Pull request by ghost**
Hi,

please review the code and tell me what you think about it.

Usage example in the shell:

``` erl-sh
159> bank:prepare(test_db, insert_t3, "INSERT INTO t3 (amount) VALUES (?)").
ok
160> bank:prepare(test_db, select_t3, "SELECT * FROM t3").                  
ok
161> bank:execute(test_db, insert_t3, [{decimal, 12000, 5}]).               
{ok,1,13}
162> bank:execute(test_db, select_t3, []).                                  
{rows,[[{<<"id">>,1},{<<"amount">>,{decimal,3333350000,5}}],
       [{<<"id">>,2},{<<"amount">>,{decimal,9999999999,5}}],
       [{<<"id">>,3},{<<"amount">>,{decimal,1000000000,5}}],
       [{<<"id">>,4},{<<"amount">>,null}],
       [{<<"id">>,5},{<<"amount">>,{decimal,-100000,5}}],
       [{<<"id">>,6},{<<"amount">>,{decimal,-1,5}}],
       [{<<"id">>,7},{<<"amount">>,{decimal,0,5}}],
       [{<<"id">>,8},{<<"amount">>,{decimal,-1,5}}],
       [{<<"id">>,9},{<<"amount">>,{decimal,-1000000,5}}],
       [{<<"id">>,10},{<<"amount">>,{decimal,-100000,5}}],
       [{<<"id">>,11},{<<"amount">>,{decimal,-10000,5}}],
       [{<<"id">>,12},{<<"amount">>,{decimal,-10000,5}}],
       [{<<"id">>,13},{<<"amount">>,{decimal,12000,5}}]]}
```

This is when you query MySQL directly:

```
mysql> select * from t3;
+----+-------------+
| id | amount      |
+----+-------------+
|  1 | 33333.50000 |
|  2 | 99999.99999 |
|  3 | 10000.00000 |
|  4 |        NULL |
|  5 |    -1.00000 |
|  6 |    -0.00001 |
|  7 |     0.00000 |
|  8 |    -0.00001 |
|  9 |   -10.00000 |
| 10 |    -1.00000 |
| 11 |    -0.10000 |
| 12 |    -0.10000 |
| 13 |     0.12000 |
+----+-------------+
13 rows in set (0.00 sec)
```

'matthias'.

### CODE REVIEW

**Comment by essen** [comment]
You'll need to rebase, I just pushed the R16B fixes.

One question immediately comes to mind: is ?MYSQL_TYPE_DECIMAL not used anymore?

**Comment by ghost** [comment]
The MySQL wire protocol compared to other databases is a big mess, if you ask me for an opinion... It was changed in version 5.1. See http://dev.mysql.com/doc/refman/5.1/en/precision-math-decimal-changes.html

I didn't implement ?MYSQL_TYPE_DECIMAL, simply because I have no older version than 5.1 to test against.

**Comment by essen** [comment]
OK cool.

**Comment by essen** [comment]
Guess you fixed it.

**Comment by ghost** [comment]
Indeed I did :-) I need to polish my knowledge about git.

File: src/bank_mysql.erl [hunk 1]
```diff
@@ -108,7 +108,7 @@
 
 %% @todo bitstring, time_tz, timestamp_tz
 -type erlang_type() :: null | integer | float
-	| date | time | timestamp | binary.
+	| date | time | timestamp | binary | decimal.
```
Comment (essen) [inline src/bank_mysql.erl]: I'd put it right after float, keep the numbers together.

File: src/bank_mysql.erl [hunk 2]
```diff
@@ -134,8 +134,9 @@
 
 -type field() :: {field, binary(), erlang_type(), mysql_type()}.
 -type remote_error() :: {remote_error, non_neg_integer(), binary(), binary()}.
+-type decimal() :: {decimal, Value :: integer(), Scale :: non_neg_integer()}.
 -type value() :: null | integer() | float()
-	| calendar:date() | calendar:time() | calendar:datetime() | binary().
+	| calendar:date() | calendar:time() | calendar:datetime() | binary() | decimal().
```
Comment (essen) [inline src/bank_mysql.erl]: Same here.

File: src/bank_mysql.erl [hunk 3]
```diff
@@ -414,6 +415,10 @@ parse_bin_row(Packet, [{field, _, integer, ?MYSQL_TYPE_LONGLONG}|Fields],
 		<< 0:1, NullRest/bits >>, Acc) ->
 	<< Value:64/little, Rest/binary >> = Packet,
 	parse_bin_row(Rest, Fields, NullRest, [Value|Acc]);
+parse_bin_row(Packet, [{field, _, decimal, ?MYSQL_TYPE_NEWDECIMAL}|Fields],
+    << 0:1, NullRest/bits >>, Acc) ->
+  {Value, Rest} = parse_decimal(Packet),
+  parse_bin_row(Rest, Fields, NullRest, [Value|Acc]);
```
Comment (essen) [inline src/bank_mysql.erl:422]: Seems incomplete. You'd need to do similar in convert_type for the sql_query function to work.

File: src/bank_mysql.erl [hunk 4]
```diff
@@ -617,6 +643,33 @@ null_map_to_mysql(Bits, Acc) ->
 	ReverseBits = << << Y:1 >> || Y <- lists:reverse(BitsList) >>,
 	<< Acc/binary, 0:Padding, ReverseBits/bits >>.
 
+decimal_to_mysql({decimal, Value, 0}) when is_integer(Value) ->
+  ValueBin = integer_to_binary(Value),
+  ValueSize = size(ValueBin),
+  <<ValueSize:8, ValueBin/binary>>;
+decimal_to_mysql({decimal, Value, Scale}) when is_integer(Value), Scale > 0 ->
+  if
+    Value >= 0 ->
```
Comment (essen) [inline src/bank_mysql.erl]: Don't use if if you can just have two clauses.

File: src/bank_mysql.erl [hunk 5]
```diff
@@ -617,6 +643,33 @@ null_map_to_mysql(Bits, Acc) ->
...
+      decimal_to_mysql(<<>>, Value, Scale);
+    true ->
+      decimal_to_mysql(<< $-:8 >>, Value * (-1), Scale)
+  end.
```
Comment (essen) [inline src/bank_mysql.erl]: Or -Value.

File: src/bank_mysql.erl [hunk 6]
```diff
@@ -617,6 +643,33 @@ null_map_to_mysql(Bits, Acc) ->
 	ReverseBits = << << Y:1 >> || Y <- lists:reverse(BitsList) >>,
 	<< Acc/binary, 0:Padding, ReverseBits/bits >>.
 
+decimal_to_mysql({decimal, Value, 0}) when is_integer(Value) ->
+  ValueBin = integer_to_binary(Value),
+  ValueSize = size(ValueBin),
+  <<ValueSize:8, ValueBin/binary>>;
```
Comment (essen) [inline src/bank_mysql.erl]: byte_size

File: src/bank_mysql.erl [hunk 7]
```diff
@@ -617,6 +643,33 @@ null_map_to_mysql(Bits, Acc) ->
...
+      decimal_to_mysql(<< $-:8 >>, Value * (-1), Scale)
+  end.
+
+decimal_to_mysql(Sign, Value, Scale) ->
+  ValueBin = integer_to_binary(Value),
+  ValueSize = size(ValueBin),
+  ValueBin2 = if
+    ValueSize =< Scale ->
+      Nulls = << << $0:8 >> || _ <- lists:seq(1, Scale - ValueSize) >>,
+      << Sign/binary, $0:8, $.:8, Nulls/binary, ValueBin/binary >>;
```
Comment (essen) [inline src/bank_mysql.erl:666]: There's better ways than using lists:seq surely.

File: src/bank_mysql.erl (hunk 6 above)
Comment (essen) [inline src/bank_mysql.erl]: byte_size everywhere

File: src/bank_mysql.erl [hunk 8]
```diff
@@ -617,6 +643,33 @@ null_map_to_mysql(Bits, Acc) ->
...
+    true ->
+      IntSize = ValueSize - Scale,
+      << _:IntSize/binary, Dec/binary >> = ValueBin,
+      << Sign/binary, ValueBin:IntSize/binary, $.:8, Dec/binary >>
+  end,
```
Comment (essen) [inline src/bank_mysql.erl:671]: Bit confusing. Match both out, then use them and not ValueBin.

**Comment by essen** [comment]
Too impatient! Anyway that should be a few things already.

**Comment by ghost** [comment]
What about tests? Wouldn't it be beneficial to add some tests?

**Comment by essen** [comment]
Yeah but we have none so far so if you want you can add them in a PR afterwards.

**Comment by ghost** [comment]
Ok, makes sense.

File: src/bank_mysql.erl (hunk 8 above)
Comment (ghost) [inline src/bank_mysql.erl:671]: What do you mean exactly? Value contains a non_neg_integer().

Comment (essen) [inline src/bank_mysql.erl:671]: << Int:InSize/binary, Dec/binary >> = ValueBin, then use Int and Dec.

Comment (ghost) [inline src/bank_mysql.erl:671]: ok, sometimes I'm blind ;-)

File: src/bank_mysql.erl (hunk 3 above)
Comment (ghost) [inline src/bank_mysql.erl:422]: has been done, see latest commit

File: src/bank_mysql.erl [hunk 9]
```diff
@@ -107,7 +107,7 @@
 -define(MYSQL_TYPE_GEOMETRY, 16#ff).
 
 %% @todo bitstring, time_tz, timestamp_tz
--type erlang_type() :: null | integer | float
+-type erlang_type() :: null | integer | float | newdecimal
```
Comment (essen) [inline src/bank_mysql.erl]: Don't think the erlang side of things would change between decimal and newdecimal, naming it decimal was fine.

File: src/bank_mysql.erl [hunk 10]
```diff
@@ -617,6 +648,35 @@ null_map_to_mysql(Bits, Acc) ->
...
+      << Int:IntSize/binary, Dec/binary >> = ValueBin,
+      << Sign/binary, Int/binary, $.:8, Dec/binary >>
+  end,
+  ValueSize2 = byte_size(ValueBin2),
+  << ValueSize2:8, ValueBin2/binary >>.
+
+fill_nulls(0, Acc) ->
+  Acc;
+fill_nulls(N, Acc) when N > 0 ->
+  fill_nulls(N - 1, << Acc/binary, $0:8 >>).
```
Comment (essen) [inline src/bank_mysql.erl]: They're 0s, not nulls. Perhaps `repeat_zeros` would be a better name?

**Comment by essen** [comment]
After that squash the commits into one and that should be good. Thanks!

**Comment by essen** [comment]
Merged! Thanks.

**Pull request merged**